crates/swc_ecma_lints/src/rules/no_dupe_args.rs

```diff
@@ -13,54 +12,65 @@ pub fn no_dupe_args() -> Box<dyn Rule> {
 #[derive(Debug, Default)]
 struct NoDupeArgs {}
 
-impl NoDupeArgs {
-    fn check(&self, param_list: Vec<(JsWord, Span)>) {
-        let mut variables_map = AHashMap::<JsWord, Span>::default();
-
-        param_list.into_iter().for_each(|(js_word, span)| {
-            if let Some(old_span) = variables_map.insert(js_word.clone(), span) {
-                HANDLER.with(|handler| {
-                    handler
-                        .struct_span_err(
-                            span,
-                            &format!(
-                                "the name `{}` is bound more than once in this parameter list",
-                                js_word
-                            ),
-                        )
-                        .span_label(old_span, "previous definition here".to_string())
-                        .span_label(span, &"used as parameter more than once".to_string())
-                        .emit();
-                });
-            }
-        });
-    }
+/// This has time complexity of O(n^2), but it's fine as the number of paramters
+/// is usually small.
+macro_rules! check {
+    ($node:expr) => {{
+        // This allocates only if there are duplicate parameters.
+        let mut done = vec![];
+
+        let mut i1 = 0;
+        for_each_binding_ident($node, |id1| {
+            i1 += 1;
+
+            let mut i2 = 0;
+            for_each_binding_ident($node, |id2| {
+                i2 += 1;
+
+                if i1 >= i2 || done.contains(&i1) {
+                    return;
+                }
+
+                done.push(i1);
+
+                if id1.span.ctxt == id2.span.ctxt && id1.sym == id2.sym {
+                    HANDLER.with(|handler| {
+                        handler
+                            .struct_span_err(
+                                id2.span,
+                                &format!(
+                                    "the name `{}` is bound more than once in this parameter list",
```

Comment:
If a binding is defined 3 times this error will be reported multiple times. That was already the case with the old implementation.

Reply:
That's handled by `done`. Actually I added `done` just to match the original behavior.